### ISSUE TITLE
json.Number support dropped

### ISSUE THREAD

**Issue by jney** [99designs/gqlgen#2521]
migrating from:

> github.com/99designs/gqlgen v0.17.22

to:

> github.com/99designs/gqlgen v0.17.24

does not support anymore my model's ID to be defined as `json.Number`:

```
cannot use res (variable of type string) as json.Number value in return statement
/Users/jney/src/ads/app/bun/graph/generated/generated.go:14618:27: cannot use v (variable of type json.Number) as string value in argument to graphql.MarshalID
```

**Comment by mstephano** [comment]
Hi, can you provide an example of model and/or schema?

**Comment by jney** [comment]
sure

go model:
```go
type Article struct {
	ID          json.Number    `json:"id"`
	// ...
}
```

gql definition:

```gql
type Article {
  id: ID!
  // ...
}
```

It still works if I don't update version. And `json.Number` are a good fit to deal with gql's `ID` that must be strings and `id` in db that are `serial`

**Comment by mstephano** [comment]
Can you show me how `ID` is defined in your `gqlgen.yml`?

**Comment by jney** [comment]
gqlgen.yml:

```
# Where are all the schema files located? globs are supported eg  src/**/*.graphqls
schema:
  - graph/*.graphqls

# Where should the generated server code go?
exec:
  filename: graph/generated/generated.go
  package: generated

# Uncomment to enable federation
# federation:
#   filename: graph/generated/federation.go
#   package: generated

# Where should any generated models go?
model:
  filename: graph/model/models_gen.go
  package: model

# Where should the resolver implementations go?
resolver:
  layout: follow-schema
  dir: graph
  package: graph

# Optional: turn on use `gqlgen:"fieldName"` tags in your models
# struct_tag: json

# Optional: turn on to use []Thing instead of []*Thing
# omit_slice_element_pointers: false

# Optional: set to speed up generation time by not performing a final validation pass.
# skip_validation: true

# gqlgen will search for any type names in the schema in these go packages
# if they match it will use them, otherwise it will generate them.
autobind:
  - "project/graph/model"

# This section declares type mapping between the GraphQL and go type systems
#
# The first line in each type will be used as defaults for resolver arguments and
# modelgen, the others will be allowed when binding to fields. Configure them to
# your liking
models:
  ID:
    model:
      - github.com/99designs/gqlgen/graphql.ID
      - github.com/99designs/gqlgen/graphql.Int
      - github.com/99designs/gqlgen/graphql.Int64
      - github.com/99designs/gqlgen/graphql.Int32
  Int:
    model:
      - github.com/99designs/gqlgen/graphql.Int
      - github.com/99designs/gqlgen/graphql.Int64
      - github.com/99designs/gqlgen/graphql.Int32
   ```

**Comment by mstephano** [comment]
@jney Thank you, I am currently investigating and will send a PR soon.

**Comment by mstephano** [comment]
@jney I ran a test with code from `v0.17.22` based on the code you provided, can you tell me that you have to implement `ID` in your `resolvers.go`? Something similar to the following:

```Go
// ID is the resolver for the id field.
func (r *userResolver) ID(ctx context.Context, obj *model.User) (string, error) {
	panic(fmt.Errorf("not implemented: ID - id"))
}
```

**Comment by jney** [comment]
something like

```go
// ID is the resolver for the id field.
func (r *someResolver) ID(ctx context.Context, obj *model.MyStruct) (string, error) {
	return string(obj.ID), nil
}
```

**Comment by mstephano** [comment]
Thank you for your confirmation. With `v0.17.24+`, that `func` will be not be needed anymore, but I am trying to find a way to fix it and not break anything else.

**Comment by mstephano** [comment]
@jney Alright, I spent some time today trying to find a solution in code, but the only conclusion I have for you for now is to create a custom `marshal/unmarshal` for `ID` like in `_examples/scalars`. Then update your `gqlgen.yml` to use it. Here is an example for you:

1. create marshal/unmarshal for `ID` in file `project/graph/model/model.go`
```Go
package model

// MarshalID is the MarshalFunc for scalar ID
func MarshalID(d json.Number) graphql.Marshaler {
	return graphql.WriterFunc(func(w io.Writer) {
		io.WriteString(w, strconv.Quote(d.String()))
	})
}

// UnmarshalID is the UnmarshalFunc for scalar ID
func UnmarshalID(v interface{}) (json.Number, error) {
	if str, ok := v.(string); ok {
		return json.Number(str), nil
	}
	return "", errors.New("ID should be string")
}
```


2. update your `gqlgen.yml` file like this:
```yml
models:
  ID:
    model: project/graph/model.ID
  Int:
    model:
      ...
```

I am sorry I could not find a better solution via the code, let me know if that works for you.

**Comment by StevenACoffman** [comment]
I think this is fixed with the latest release v0.17.27 

@jney Can you please verify that this issue is resolved?
